server.go

```diff
@@ -191,17 +194,25 @@ func (srv *Server) Serve(l net.Listener) error {
 	}
 }
 
-func (srv *Server) handleConn(conn net.Conn) {
+func (srv *Server) handleConn(newConn net.Conn) {
 	if srv.ConnCallback != nil {
-		cbConn := srv.ConnCallback(conn)
+		cbConn := srv.ConnCallback(newConn)
 		if cbConn == nil {
-			conn.Close()
+			newConn.Close()
 			return
 		}
-		conn = cbConn
+		newConn = cbConn
+	}
+	ctx, cancel := newContext(srv)
+	conn := &serverConn{
+		Conn:          newConn,
+		idleTimeout:   srv.IdleTimeout,
+		closeCanceler: cancel,
```

Comment:
The point was to give the connection the `cancel` not just for when it's closed but when `net.Error` is returned from a Read or Write.